metabase/metabase
# Admin logs layout fix & perf improvements

Close #39025

### Description

**Auto-follow bug**
There was a layout issue where if a single log was taller than the contain for the logs it would make the container grow taller than what would be visible to the user. This affected the calculation of if the users was scrolled to the bottom and would disable the auto-follow feature.

**User scroll position bug**
The logs page was not preserving the user's scroll position in some cases when logs would be updated. Upon inspecting our internal MB instance, it appears this was due to request times taking longer than our poll duration, causing state updates to happen out of order. This would remove logs from the list and thus change the relative position of other logs.

**Performance issues**
In our internal MB instance, payload sizes were coming in ~200kb-800kb gziped / ~2MB-5MB uncompressed. Given we combine logs from more than one machine, the internal list of logs  combined to ~8MB. Since we are polling for new updates, we were taking our 8MB list of logs and trying to merge / 2-5MB on top of it. Between our merging implementation and usage of map/filter in our Logs component's render method, we were making several copies of this in memory every second. This results in large freezes in the UI. Here you can see setState alone taking almost 3 second to update:
<img width="421" alt="Screenshot 2024-02-22 at 4 59 36 PM" src="https://github.com/metabase/metabase/assets/7104357/de5f001e-ea05-43e5-a95a-3b16d2d43bbe">

**The fix**

1. For the auto-follow bug - it was a matter of using flex on the container component for force the content to shrink to the expected size.
2. User scroll position - I have updated the UI to skip polling for new logs if a previous request is still in flight.
3. Performance issues
  a. skipping requests helps
  b. updated mergeLogs function to be more performant (3 sorts w/ 3 copies reduced to 1, _.uniq was slow so replaced with hand rolled dedup implementation)
  c. skip state merging / state updates if latest log in the most recent result is already present in our state
  d. moment date formatting on large logs sets was slow, moving to dayjs and caching the last result reduced date significantly

### How to verify

- go to admin logs page
- generate a large number of logs (for internal reviewers, you can borrow the response from stats)
- override your response body in dev tools with your large logs set ([how to do this in Chrome](https://developer.chrome.com/docs/devtools/overrides#set-up))
- your page should no longer freeze

### Checklist
